aiohttp/http_parser.py

```diff
@@ -757,16 +749,8 @@ def __init__(
             if self._length == 0:
                 real_payload.feed_eof()
                 self.done = True
-        else:
-            if readall and code != 204:
-                self._type = ParseState.PARSE_UNTIL_EOF
-            elif method in ("PUT", "POST"):
-                internal_logger.warning(  # pragma: no cover
-                    "Content-Length or Transfer-Encoding header is required"
-                )
-                self._type = ParseState.PARSE_NONE
-                real_payload.feed_eof()
-                self.done = True
+        elif code != 204:
```

Comment:
I'm not sure we can get here if the code is 204

Reply:
204 will always be empty body and the only place that can get in here is 

https://github.com/aio-libs/aiohttp/pull/8096/files#diff-197153342549adb4b25580f8b842876841eec42dc48ae12cb72ba2ae4f7ced10R380

AFAICT

Reply:
I think if it's CONNECT and 204 it might reach here, which is why I wasn't sure about removing the check. Not sure if that's actually relevant or not...

Reply:
Maybe we should stop CONNECT getting in here too? In which case, maybe we can refactor the code to avoid the duplicate `not empty_body` checks.

Reply:
https://www.rfc-editor.org/rfc/rfc9110#section-9.3.6-7
> Any [2xx (Successful)](https://www.rfc-editor.org/rfc/rfc9110#status.2xx) response indicates that the sender (and all inbound proxies) will switch to tunnel mode immediately after the response header section; data received after that header section is from the server identified by the request target. Any response other than a successful response indicates that the tunnel has not yet been formed.

> A CONNECT request message does not have content. The interpretation of data sent after the header section of the CONNECT request message is specific to the version of HTTP in use.


I think if I'm reading that correctly, then we should read the body regardless of whether it's 204 or not. So, I think we want to remove the check and always read?

Reply:
Anything after the response is going to be the proxied socket.  After the response headers I think we are only forwarding the stream.  I haven't dug through the paths to figure out how that works yet, but it seems like we shouldn't read the body and at that point its effectively forwarding read/writes of a stream?

Reply:
So, maybe the body is supposed to be read when it's not a 2xx response? I'm not clear why this code seems to always try to read the response when the method is CONNECT..

Reply:
I was a bit worried that someone has a use case for doing CONNECT and than speaking some completely different protocol. AFAICT the only use case for CONNECT is for proxies. 

https://github.com/aio-libs/aiohttp/blob/0ec65c0f4dc08d027f659256b09ae9cff10ab404/aiohttp/connector.py#L1263

I think this change looks good as we are never going to get in to `HttpPayloadParser` on empty body responses so we only have to worry about it on `METH_CONNECT` and it seems like the logic remains unchanged.

Reply:
Plot twist: I did some Git paleontology and it turned out that the patch implementing this bit of logic is the same commit that marked the exactly same proxy tests that I mentioned earlier in this PR (https://github.com/aio-libs/aiohttp/pull/8096#issuecomment-1914876703) with xfail. See https://github.com/aio-libs/aiohttp/commit/b26702449142733caa0e026b331381bf5bee4f01#diff-be15bec475f6c781d5d3e4413c19d0e24ccaf616b8a2495c6824fc3a16d139f5R270.

Does this mean that the original implementer wasn't sure the implementation was correct 7 years ago? Will those tests miraculously start passing if this is removed?

Reply:
> Will those tests miraculously start passing if this is removed?

I don't think we're making any real change to behaviour, so no, I don't think they'll start passing. I'd note that someone renamed the test to `xtest` though, so they are currently not running at all. The test you linked to still fails when I run locally with: `ClientConnectorError: Cannot connect to host www.google.com.ua:443 ssl:default [None]`